apis/ops/v1alpha1/singlestore_ops_types.go

```diff
@@ -0,0 +1,124 @@
+import (
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	ResourceCodeSinglestoreOpsRequest     = "sdbops"
+	ResourceKindSinglestoreOpsRequest     = "SinglestoreOpsRequest"
+	ResourceSingularSinglestoreOpsRequest = "singlestoreopsrequest"
+	ResourcePluralSinglestoreOpsRequest   = "singlestoreopsrequests"
+)
+
+// SinglestoreOpsRequest defines a Singlestore DBA operation.
+
+// +genclient
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=singlestoreopsrequests,singular=singlestoreopsrequest,shortName=sdbops,categories={datastore,kubedb,appscode}
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Type",type="string",JSONPath=".spec.type"
+// +kubebuilder:printcolumn:name="Status",type="string",JSONPath=".status.phase"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+type SinglestoreOpsRequest struct {
+	metav1.TypeMeta   `json:",inline,omitempty"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              SinglestoreOpsRequestSpec `json:"spec,omitempty"`
+	Status            OpsRequestStatus          `json:"status,omitempty"`
+}
+
+// SinglestoreOpsRequestSpec is the spec for SinglestoreOpsRequest
+type SinglestoreOpsRequestSpec struct {
+	// Specifies the Singlestore reference
+	DatabaseRef core.LocalObjectReference `json:"databaseRef"`
+	// Specifies the ops request type: UpdateVersion, HorizontalScaling, VerticalScaling etc.
+	Type SinglestoreOpsRequestType `json:"type"`
+	// Specifies information necessary for vertical scaling
+	VerticalScaling *SinglestoreVerticalScalingSpec `json:"verticalScaling,omitempty"`
+	// Specifies information necessary for volume expansion
+	VolumeExpansion *SinglestoreVolumeExpansionSpec `json:"volumeExpansion,omitempty"`
+	// Specifies information necessary for restarting database
+	Restart *RestartSpec `json:"restart,omitempty"`
+	// Specifies information necessary for custom configuration of Singlestore
+	Configuration *SinglestoreCustomConfigurationSpec `json:"configuration,omitempty"`
+	// Timeout for each step of the ops request in second. If a step doesn't finish within the specified timeout, the ops request will result in failure.
+	Timeout *metav1.Duration `json:"timeout,omitempty"`
+	// ApplyOption is to control the execution of OpsRequest depending on the database state.
+	// +kubebuilder:default="IfReady"
+	Apply ApplyOption `json:"apply,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=VerticalScaling;VolumeExpansion;Restart;Configuration
+// ENUM(VerticalScaling, VolumeExpansion, Restart, Configuration)
+type SinglestoreOpsRequestType string
+
+// SinglestoreReplicaReadinessCriteria is the criteria for checking readiness of a Singlestore pod
+// after updating, horizontal scaling etc.
+type SinglestoreReplicaReadinessCriteria struct{}
+
```

Comment:
I don't see any usage of it. Remove

Reply:
I used it in autoscaler spec. Other db have ReadinessCriteria. in my case, I've set the replica readiness criteria empty now. Maybe in the future, will add the criteria like others db. 

type SinglestoreAutoscalerSpec struct {
	DatabaseRef *core.LocalObjectReference `json:"databaseRef"`

	// This field will be used to control the behaviour of ops-manager
	OpsRequestOptions *SinglestoreOpsRequestOptions `json:"opsRequestOptions,omitempty"`

	Compute *SinglestoreComputeAutoscalerSpec `json:"compute,omitempty"`
	Storage *SinglestoreStorageAutoscalerSpec `json:"storage,omitempty"`
}

type SinglestoreOpsRequestOptions struct {
	// Specifies the Readiness Criteria
	ReadinessCriteria *opsapi.SinglestoreReplicaReadinessCriteria `json:"readinessCriteria,omitempty"`

	// Timeout for each step of the ops request in second. If a step doesn't finish within the specified timeout, the ops request will result in failure.
	Timeout *metav1.Duration `json:"timeout,omitempty"`

	// ApplyOption is to control the execution of OpsRequest depending on the database state.
	// +kubebuilder:default="IfReady"
	Apply opsapi.ApplyOption `json:"apply,omitempty"`
}